arp-speaker/main.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+
+	"go.universe.tf/metallb/internal/allocator"
+	"go.universe.tf/metallb/internal/arp"
+	"go.universe.tf/metallb/internal/config"
+	"go.universe.tf/metallb/internal/k8s"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"k8s.io/api/core/v1"
+)
+
+type controller struct {
+	myIP   net.IP
+	myNode string
+
+	config *config.Config
+	ips    *allocator.Allocator
+	ann    *arp.Announce
+
+	// Metrics
+	announcing *prometheus.GaugeVec
+}
+
+func (c *controller) SetBalancer(name string, svc *v1.Service, eps *v1.Endpoints) error {
+	if svc == nil {
+		return c.deleteBalancer(name, "service deleted")
+	}
+
+	if svc.Spec.Type != "LoadBalancer" {
+		return nil
+	}
+
+	glog.Infof("%s: start update", name)
+	defer glog.Infof("%s: end update", name)
+
+	if c.config == nil {
+		glog.Infof("%s: skipped, waiting for config", name)
+		return nil
+	}
+
+	if len(svc.Status.LoadBalancer.Ingress) != 1 {
+		glog.Infof("%s: no IP allocated by controller", name)
+		return c.deleteBalancer(name, "no IP allocated by controller")
+	}
+
+	// Should we advertise? Yes, if externalTrafficPolicy is Cluster,
```

Comment:
Note for future changes: the arp logic needs to run leader election, and only 1 node should advertise. Sadly ARP doesn't really support LB, and things won't work right if there's >1 ARP responder for a VIP.

But, this is just a copypasta  import for now, no need to change it for this CL.

Reply:
I was hoping to postpone this until every thing is more polished, although leader-election in a k8s might be trivial to implement.

Reply:
There is https://godoc.org/k8s.io/client-go/tools/leaderelection . It's a library that uses k8s Endpoints objects to implement leader election, similar to how Chubby does it (try to atomically write the leader's IP to the file, if you succeed, you are the leader).

It's fine to do this whenever you want :)